Bugfix 14631: Enum with comma produces incorrect search dropdown

WellingtonBraga opened the pull request:
### Description

Signed-off-by: Wellington Braga <wellington.braga.inatel@gmail.com>

When a comma is used withing the value of an ENUM field the search dropdown incorrectly splits it.

The bug occurs due a wrong split logic. The split function was assuming just a single quote as delimiter. This commit alters the delimiter to a single quote followed by a comma.

Fixes #14631 

Before submitting pull request, please review the following checklist:

- [x] Make sure you have read our [CONTRIBUTING.md](https://github.com/phpmyadmin/phpmyadmin/blob/master/CONTRIBUTING.md) document.
- [x] Make sure you are making a pull request against the correct branch. For example, for bug fixes in a released version use the corresponding QA branch and for new features use the _master_ branch. If you have a doubt, you can ask as a comment in the bug report or on the mailing list.
- [x] Every commit has proper `Signed-off-by` line as described in our [DCO](https://github.com/phpmyadmin/phpmyadmin/blob/master/DCO). This ensures that the work you're submitting is your own creation.
- [x] Every commit has a descriptive commit message.
- [x] Every commit is needed on its own, if you have just minor fixes to previous commits, you can squash them.
- [x] Any new functionality is covered by tests.

MauricioFauth reviewed (changes requested):
Hello, @WellingtonBraga. Thank you for your contribution.
You should use `QA_4_8` as the base branch, since this issue is a bug fix.

This PR does not fully fix the issue, because a space is added after the comma. This happens in the following lines:
https://github.com/phpmyadmin/phpmyadmin/blob/69dcf2caf8107ab0f4fc33fceb0cf4ce625477dd/libraries/classes/Controllers/Table/TableSearchController.php#L145-L148

WellingtonBraga commented on templates/table/search/input_box.twig:
```diff
@@ -44,7 +44,7 @@
     or (column_type starts with 'set' and in_zoom_search_edit) %}
     {% set in_zoom_search_edit = false %}
-    {% set value = column_type|e|slice(5, -1)|replace({'&#039;': ''})|split(', ') %}
+    {% set value = column_type|trim|slice(6, -2)|split('\', \'') %}
```
I went through the code to see the most appropriated way to apply parseEnumSetValues. I found out some implications that I'd like to point.

- The current implementation of this page, parses the enum values directly on the template. If we want to use the method you mentioned we should create a kind of new filter just for that, which could be much more effort than this bug actually needs.

I saw it here:
https://stackoverflow.com/questions/32920217/how-to-call-static-function-in-symfony2-twig-template

We could call this method in controller and got an array with the values from that. I could use something like this:

`$this->_columnTypes[] = strpos($type, 'enum') !== false ? Util::parseEnumSetValues($type) : $type;`

But still we'd have more effort than we needs since, we then gonna need to redesign the template to mirror this change. The template is expecting a string (enum('op1','op2')) and give it a array implies in some refactor on the entire logic of printing enum options

About the space you mentioned before, it is already consider in the new split delimiter I am using.

Maybe I am not seeing what you are seeing. In this case, I would be glad if you could clarify them to me. 

I'm looking forward to contribute more and I hope to have more excelent insights like yours

mauriciofauth replied:
You can register _parseEnumSetValues_ as a Twig function in PhpMyAdmin\Twig\UtilExtension, and then pass `column_type` as a parameter.

About the space issue, if you have `enum('foo,bar','baz')`, spaces are being added after commas, like this: `enum('foo, bar', 'baz')`. We don't want to change the enum values.

You can see how enum and set type are handled, here:
https://github.com/phpmyadmin/phpmyadmin/blob/f44c3c6939865b5aa96b1c0412117df5aa5c6321/libraries/classes/Util.php#L2794-L2798

WellingtonBraga replied:
Hi Mauricio..
I prepared a new solution based on your insights. Could you please re-check it?
Thanks again for that. I learned a little bit more about the project and how it is structured.

Hope to contribute more.

Thanks in advance.

WellingtonBraga commented on templates/table/search/input_box.twig:
```diff
@@ -69,11 +69,11 @@
         {% else %}
             <option value="{{ value[i]|raw }}">
-                {{ value[i]|raw }}
+                {{ value[i]|raw|e }}
```
Done.

WellingtonBraga commented on templates/table/search/input_box.twig:
```diff
@@ -69,11 +69,11 @@
             and value[i] in criteria_values[column_index] %}
             <option value="{{ value[i]|raw }}" selected>
-                {{ value[i]|raw }}
+                {{ value[i]|raw|e }}
```
Done.

Commit:
Bugfix 14631: Enum with comma produces incorrect search dropdown

When a comma is used withing the value of an ENUM field the search dropdown incorrectly splits it.

The bug occurs due a wrong split logic. The split function was assuming just a single quote as delimiter. This commit alters the delimiter to a single quote followed by an comma.

Signed-off-by: Wellington Braga <wjbraga@artit.com.br>

mauriciofauth commented on templates/table/search/input_box.twig:
```diff
@@ -68,12 +68,12 @@
-            <option value="{{ value[i]|raw }}">
-                {{ value[i]|raw }}
+            <option value="{{ value[i]|e }}">
+                {{ value[i]|e }}
```
All variables in Twig templates are automatically escaped. So you don't need explicitly add the escape filter.

However, as you didn't pass `false` to `Util_parseEnumSetValues` second argument, parseEnumSetValues are already escaping the values, so you need to use the `raw` filter here instead of `e` to not double escape the values.

MauricioFauth reviewed (changes requested):
There is still an issue with a space been added after all commas.
With we have a column with the type `enum('foo,bar')`, we don't want to add a space after the comma, because the comma is part of the enum value.
